Haiku port

mmuman opened the pull request:
This adds [Haiku](https://www.haiku-os.org/) support using the native BKeyStore API.
It seems to work but there are still some concerns.
At least the test code needs a hack to avoid an "invalid team" error, but that's probably a bug in QCoreApplication, or a misuse (maybe it needs to be started in some way).

ogoffart commented on keychain_p.h:
```diff
@@ -15,7 +15,7 @@
 #include <QQueue>
 
-#if defined(Q_OS_UNIX) && !defined(Q_OS_DARWIN) && !defined(Q_OS_ANDROID)
+#if defined(Q_OS_UNIX) && !defined(Q_OS_DARWIN) && !defined(Q_OS_ANDROID) && !defined(Q_OS_HAIKU)
```
Maybe worth doing a  `#define KEYCHAIN_XDG`  or something like that (maybe KEYCHAIN_DBUS).  Then we can  use that ifdef instead of repeating the same condition all over the place

mmuman replied:
Yes, I didn't suggest it because the line didn't wrap yet but it's cleaner.
We can even have CMake do it.

mmuman replied:
what about !KEYCHAIN_HAVE_NATIVE ?

ogoffart replied:
In a way, secret service is the native way,  so i guess that name is not good.   XDG is the name used for the typicall desktop environment on X11.

mmuman replied:
Technically XDG is a set of standards from freedesktop.org , most of which written with X11 in mind (wonder what that will give with Wayland:D), but some of those could also possibly be supported in Haiku.
So I'm not sure it really fits.

ogoffart commented on keychain_haiku.cpp:
```diff
@@ -0,0 +1,129 @@
+                                      q->key().toUtf8().constData(),
+                                      false, password);
+
+    data = QByteArray((char*)password.Data(), password.DataLength());
```
Don't use C casts. use `static_cast<>`  (or `reinterpret_cast<>`)

mmuman replied:
Ok, I copy-pasted this line from the win code :D

ogoffart commented on keychain_haiku.cpp:
```diff
@@ -0,0 +1,129 @@
+        q->key().toUtf8().constData());
+    status_t result;
+
+    // XXX: is it a string or a blob??
```
That depends on JobPrivate::mode

mmuman replied:
I don't know our API much on this topic, and it hasn't been tested much yet, but I think the BPasswordKey class expects a string, while the base class is more flexible. Maybe I should just use the base class and set the fields separately.

ogoffart commented on keychain_haiku.cpp:
```diff
@@ -0,0 +1,129 @@
+/******************************************************************************
+ *   Copyright (C) 2011-2015 Frank Osterfeld <frank.osterfeld@gmail.com>      *
```
Maybe you want to change that line.

mmuman replied:
Added a fixup commit, I'll rebase later on.

ogoffart commented on testclient.cpp:
```diff
@@ -23,6 +26,10 @@ static int printUsage() {
     QCoreApplication app( argc, argv );
+#ifdef __HAIKU__
+    // XXX: for some reason the QApp object doesn't create a BApp
+    BApplication bapp("application/x-vnd.QtKeychain");
```
How do other Qt application do it?  Maybe it only works with QApplication or QGuiApplication?
If that's the case, then better change to QApplication for haiku.

mmuman replied:
I'll try with it.
Either that or it is just created later on if we actually start it.
In Haiku as well sometimes we create the app object without calling Run() on it because it's not required for what we need.

ogoffart commented on testclient.cpp:
```diff
@@ -8,6 +8,9 @@
  *****************************************************************************/
 #include <QCoreApplication>
 #include <QStringList>
+#ifdef __HAIKU__
```
`Q_OS_HAIKU`

mmuman replied:
Right, ideally we can fix and drop this anyway.

mmuman commented:
Well, I ported countless things to BeOS, then Haiku, over the years, and it did happen that BeOS code was dropped due to lack of maintainer.
Only few projects actually really complained on patches (like, irssi saying "we don't care about your antique OS", despite them having support for HP/UX in their configure, but LOL. And even irssi ended up accepting most patches as they actually cleaned up their code as well).

Besides, things have changed much in recent years. We now have a proper ports tree, [HaikuPorts](https://github.com/haikuports/haikuports/), and even bots to build packages automatically.
I ported this because I need it for the NextCloud/ownCloud desktop client, and it's likely it'll be used.
I can leave the patchset in HaikuPorts anyway, but I'd rather have it upstream.

However, there aren't that many people maintaining recipes, so we don't always update them for every release of everything. Also, the shear amount of github notifications will likely hide any release notice I could get. Maybe with a thunderbird filter… But I'd rather try to automate things.

Now, it's not like there are that many Haiku-specific stuff around, it's quite self-contained, so even if the next release doesn't build on Haiku it shouldn't hurt other platforms. I'm thinking about a script to check for new release on the recipes for things on github. It wouldn't ensure the release builds as it'd be too late but at least make sure we fix it for the next one. Another option would be to have a recipe pointing to the git HEAD and have it checked regularly, but I'll have to dedicate a machine for CI.

What do you think would be acceptable?

mmuman commented:
Hmm I wanted to test things on linux to make sure I don't break it, but for some reason cmake doesn't detect Qt5 on my Debian Sid :-(

mmuman commented:
Btw, I always get some .ts files in translations/ that are modified by the build (version set to 2.1 instead of 2.0 and lots of changes). Are those generated or just overwritten by bad behaving tools?

mmuman commented:
Ok, force-pushed a cleaned-up version.
I addressed all the issues on the haiku code, the only thing left is the test hack, as I couldn't get it to work with QApplication yet, I'll have a look again.

ogoffart commented:
> What do you think would be acceptable?

I personally do think this patch is acceptable since the haiku part is well isolated.
I'm just saying we won't be maintaining it, and that it might bitrot in the future.

mmuman commented:
Okay, rebased and tested on Debian, it still works there.
Seems I was missing some packages… qtbase5-dev extra-cmake-modules qttools5-dev

Last pending issue is the QCoreApplication workaround I think.

mmuman commented:
So, I figured I would just check for a valid BApplication object and create one if needed anyway. It shouldn't matter if someday QCoreApplication actually creates one, be_app won't be NULL anymore then and we'll skip the code.

I think we're ready now.

mmuman commented:
Just rebased.

mmuman commented:
Rebased again, builds fine.

mmuman commented:
I just rebased again. I don't think there's any reason left not to merge.

Btw, I just added an option to our [haikuporter](https://github.com/haikuports/haikuporter/) tool to check for newer releases on github, so we can now automate detecting new release and better react by updating the recipes and fixing breakages. This should address your concern about bitrot.

frankosterfeld commented on keychain_haiku.cpp:
```diff
@@ -9,10 +9,64 @@
 
+class AutoApp {
+public:
+	AutoApp();
```
This commit seems to use tabs, please use four spaces instead.

frankosterfeld replied:
Code looks fine otherwise.

mmuman replied:
Ah yes, that's the default for Haiku…
Do you want it squashed?

frankosterfeld commented:
> That's what I meant by squashed ;-)

Sorry, I missed that comment

mmuman commented:
Ok, I squashed the Haiku port into a single commit.

Btw, doing a `make clean` removes the translation files, then running `make` gives:
> lupdate error: List file '/work/qtkeychain/qtkeychain/build/CMakeFiles/qtkeychain_zh.ts_lst_file' is not readable.

It's really annoying also when you build that it changes them, I always end up having to git checkout to clean the working directory…

frankosterfeld commented:
Thanks!
